File: 098. Validate Binary Search Tree/isValidBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    var traverse = (root,low,high) => {
        if (!root) {
            return true
        }
        else if (low >= root.val) { //can't use !(low < root.val <high)
            return false
        }
        else if (high <= root.val) {
            return false
        }
        else {
            return traverse(root.left,low,root.val) && traverse(root.right,root.val,high)
        }
    }
    return traverse(root,low=-Infinity,high=Infinity)
};

// Runtime 78 ms / Memory 46.5 MB
// debug: https://jsfiddle.net/

File: 098. Validate Binary Search Tree/isValidBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        long long int boundaryCondition_L = LLONG_MIN; //less than min spec
        long long int boundaryCondition_R = LLONG_MAX; //large than max spec
        return DFS(root, boundaryCondition_L, boundaryCondition_R);
    }

    bool DFS(TreeNode* node, long long int BC_L, long long int BC_R) {
        if (!node) { return true; }
        else if (node->val <= BC_L ) { return false; }
        else if (node->val >= BC_R ) { return false; }
        else {
            bool L = DFS(node->left, BC_L, node->val);
            bool R = DFS(node->right, node->val, BC_R);
            return (L && R);
        }
    }
};